src/test/kotlin/org/jetbrains/kotlinx/jupyter/test/repl/ReplTests.kt

```diff
@@ -441,3 +442,217 @@ class ReplTests : AbstractSingleReplTest() {
+        varsState.forEach {
+            strState[it.key] = it.value.stringValue ?: return@forEach
+        }
+
+        val returnedState = res.metadata.variablesMap
+        assertEquals(strState, returnedState)
+        assertEquals("1", varsState["x"]!!.stringValue)
+        assertEquals("abc", varsState["y"]!!.stringValue)
+        assertEquals("1", varsState["z"]!!.stringValue)
+    }
+
+    @Test
+    fun testVarsCaptureSeparateCells() {
+        eval(
+            """
+            val x = 1 
+            val y = "abc"
+            val z = x
+            """.trimIndent()
+        )
+        val varsState = repl.notebook.variablesMap
+        assertTrue(varsState.isNotEmpty())
+        eval(
+            """
+            val x = "abc" 
+            var y = 123
+            val z = x
+            """.trimIndent(),
+            jupyterId = 1
+        )
+        assertTrue(varsState.isNotEmpty())
+        assertEquals(3, varsState.size)
+        assertEquals("abc", varsState["x"]!!.stringValue)
+        assertEquals("123", varsState["y"]!!.stringValue)
+        assertEquals("abc", varsState["z"]!!.stringValue)
+
+        eval(
+            """
+            val x = 1024 
+            y += 123
+            """.trimIndent(),
+            jupyterId = 2
+        )
+
+        assertTrue(varsState.isNotEmpty())
+        assertEquals(3, varsState.size)
+        assertEquals("1024", varsState["x"]!!.stringValue)
+        assertEquals("${123 * 2}", varsState["y"]!!.stringValue)
+        assertEquals("abc", varsState["z"]!!.stringValue)
+    }
+
+    @Test
+    fun testVarsUsageConsistency() {
+        eval("3+2")
+        val state = repl.notebook.usageMap
+        assertTrue(state.values.size == 1)
+        assertTrue(state.values.first().isEmpty())
+        val setOfNextCell = setOf<String>()
+        assertEquals(state.values.first(), setOfNextCell)
+    }
+
+    @Test
+    fun testVarsDefsUsage() {
+        eval(
+            """
+            val x = 1
+            val z = "abcd"
+            var f = 47
+            """.trimIndent()
+        )
+        val state = repl.notebook.usageMap
+        assertTrue(state.isNotEmpty())
+        assertTrue(state.values.first().isNotEmpty())
+        val setOfCell = setOf("z", "f", "x")
+        assertTrue(state.containsValue(setOfCell))
+    }
+
+    @Test
+    fun testVarsDefNRefUsage() {
+        eval(
+            """
+            val x = "abcd"
+            var f = 47
+            """.trimIndent()
+        )
+        val state = repl.notebook.usageMap
+        assertTrue(state.isNotEmpty())
+        eval(
+            """
+            val z = 1
+            f += f
+            """.trimIndent()
+        )
+        assertTrue(state.isNotEmpty())
+
+        val setOfCell = setOf("z", "f", "x")
+        assertTrue(state.containsValue(setOfCell))
+    }
+
+    @Test
+    fun testSeparateDefsUsage() {
+        eval(
+            """
+            val x = "abcd"
```

Comment:
Please make the tests for private/protected/internal variables:
```
private val x = 5
```
We should be sure that such declarations do not crash the notebook

Reply:
Improved realisation. All working correctly, but we have a limitation such that it's not possible to access any fields with non-default modificators in separate cells to modify. 
For instance, we have 2 cells. 
One: 
`private val x = 5 `
Another:
```
private var z = 47
z += x
```
Compile-time error, I believe it's expected